lib/auth/methods.go

```diff
@@ -136,6 +127,30 @@ func (s *Server) AuthenticateUser(ctx context.Context, req AuthenticateUserReque
 		},
 		Method: events.LoginMethodLocal,
 	}
+	defer func() {
+		if err := s.emitter.EmitAuditEvent(s.closeCtx, event); err != nil {
+			log.WithError(err).Warn("Failed to emit login event.")
+		}
+	}()
+
+	mfaDev, actualUsername, err := s.authenticateUser(ctx, req)
+	if err != nil {
+		event.Code = events.UserLocalLoginFailureCode
+		event.Status.Success = false
+		event.Status.Error = err.Error()
+		return nil, trace.Wrap(err)
+	}
+
+	switch {
+	case username != "" && actualUsername != "" && username != actualUsername:
+		log.Warnf("Authenticate user mismatch (%q vs %q). Using request user (%q)", username, actualUsername, username)
+	case username == "" && actualUsername != "":
+		log.Debugf("User %q authenticated via passwordless", actualUsername)
+		username = actualUsername
+	}
+
+	event.Code = events.UserLocalLoginCode
```

Comment:
Looks like half of this function's logic now is to build fields for the audit events. Can it be factored out into a separate method so the function looks like this?
```go
func (s *Server) AuthenticateUser() xxx {
    xxx, err := s.authenticateUser(xxx)
    // handle err

    xxx := s.emitAuthAuditEvent(yyy)
    // ...
}
```